mongo/options.go

```diff
@@ -0,0 +1,28 @@
+// Copyright (C) MongoDB, Inc. 2023-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package mongo
+
+import (
+	"go.mongodb.org/mongo-driver/internal/mongoutil"
+)
+
+// Options is an interface that wraps a method to return a list of setter
+// functions that can set a generic arguments type.
+type Options[T mongoutil.Args] interface {
```

Comment:
Is there a risk to using `T any` instead of `T mongoutil.Args`?

Reply:
Although, functionally, `T any` fits here as well, I think it's better to restrict the types. The `mongoutil.Args` type also implies the expected usage. The only issue is that we need to maintain the `Args` list.

Reply:
I'm concerned about the cost of maintaining that exhaustive list and also of using something defined in an internal package in the `options` API. The usage of `options.Options` in the `mongo` and `options` APIs seems pretty clear already.

Reply:
No matter what, we would need a compile-time check that options implement `ArgsSetters`. `mongoutil.Args` serves this purpose under the proposed paradigm, as well as providing type safety to the `mongo.Options` type. 

In terms of maintaining the list, options are part of stable API so this list could only grow within the scope of 2.x.

Unfortunately, due to dependency cycles we can't define this list in the mongo package. However, it doesn't really matter: if an embedded type (WLOG consider options.ClientArgs) is removed from `mongoutil.Args`, then you will get a compile-time failure akin to the following:

```
"go.mongodb.org/mongo-driver/mongo/options".ClientArgs does not satisfy mongoutil.Args
```

The _only_ way you wouldn't get this failure is if (1) the underlying options are not used [in which case removal doesn't matter], or (2) the options have been removed in all instance [a breaking change].

IMO the benefit of type-safety outweighs the need to extend this list when new option requirements arise.

Reply:
I was actually thinking about when we need to add new `Options` types, which would require adding a type to the `mongoutil.Args` interface. For example, if we add a new `Collection` method `coll.Foo`, we'd probably also add `FooOptions` and `FooArgs`, which will need to implement `Options`. We'd need to add `FooArgs` to `mongoutil.Args`. Could that change break any user code?

Another consideration is users who want to write functions that operate on `Options`. For example, if a user wanted to pass `*Args` types directly to collection methods (like we do in our spec tests), they might define a generic type like:
```go
type argsAsOptions[T mongoutil.Args] struct {
	args *T
}

func (o *argsAsOptions[T]) ArgsSetters() []func(*T) error {
	return []func(*T) error{
		func(args *T) error {
			*args = *o.args
			return nil
		},
	}
}

func main() {
	coll.FindOne(
		context.Background(),
		bson.D{},
		&argsAsOptions[options.FindOneArgs]{
			args: &options.FindOneArgs{
				Comment: "blah",
			},
		},
	)
}
```

However, a user can't reference `mongoutil.Args` from outside the Go Driver module, so they would have to define a type constraint that is a copy of `mongoutil.Args`:
```go
type myArgs interface {
	options.AggregateArgs | options.BucketArgs | options.BulkWriteArgs |
	// ...
}

type argsAsOptions[T myArgs] struct {
	// ...
```
That example is inspired by a public code search for users defining `ClientOptions` literals instead of using the builder API (see [here](https://sourcegraph.com/search?q=context:global+lang:Go++-repo:%5Egithub%5C.com/mongodb/mongo-go-driver%24+repo:has.content%28go.mongodb.org/mongo-driver%29+-file:%28%5E%7C/%29vendor/+%26options.ClientOptions&patternType=standard&case=yes&sm=0&groupBy=repo)). Is that a use case we want to support?

Concerning type safety, is there a problematic case that using the `mongoutils.Args` type constraint prevents?

Reply:
> I was actually thinking about when we need to add new Options types, which would require adding a type to the mongoutil.Args interface. 

We do this as a compile-time check, it's essentially this:

```go
var _ mongoutil.Args = &FindOneArgs{}
```

Removing any arg struct (that is actually being used somewhere in the code) from `mongoutil.Args` will result in the following error [WLOG]:

```
"go.mongodb.org/mongo-driver/mongo/options".ClientArgs does not satisfy `mongoutil.Args`
```

> Could that change break any user code?

Users don't need to be able to access or copy `mongoutil.Args`. Although it's unclear to me why you would want to do this, if a user is trying to genericize their own options they can create a type that implements a subset of `mongoutil.Args`, e.g.:

```go
type opts interface {
	options.FindOneArgs | options.InsertOneArgs | SomeRandomStruct
}

type argsAsOptions[T opts] struct {
	args *T
}
```

In general, I would expect users to define custom types without generics:

```go
type argsAsOptions struct {
	args *options.FindOneArgs
}

func (o *argsAsOptions) ArgsSetters() []func(*options.FindOneArgs) error {
	return []func(*options.FindOneArgs) error{
		func(args *options.FindOneArgs) error {
			*args = *o.args

			return nil
		},
	}
}

func main() {
	// ...

	coll.FindOne(
		context.Background(),
		bson.D{},
		&argsAsOptions{
			args: &options.FindOneArgs{
				Comment: "blah",
			},
		},
	)
}
```

> That example is inspired by a public code search for users defining ClientOptions literals instead of using the builder API (see [here](https://sourcegraph.com/search?q=context:global+lang:Go++-repo:%5Egithub%5C.com/mongodb/mongo-go-driver%24+repo:has.content%28go.mongodb.org/mongo-driver%29+-file:%28%5E%7C/%29vendor/+%26options.ClientOptions&patternType=standard&case=yes&sm=0&groupBy=repo)). Is that a use case we want to support?

I would also expect users to define struct literals by appending the `Opts` field:

```go
opts := options.Client().ApplyURI("mongodb://localhost:27017")

opts.Opts = append(opts.Opts, func(args *options.ClientArgs) error {
	appName := "myApp"

	*args = options.ClientArgs{
		AppName: &appName,
	}

	return nil
})

client, err := mongo.Connect(opts)
if err != nil {
	panic(err)
}
```

> Concerning type safety, is there a problematic case that using the mongoutils.Args type constraint prevents?

Probably not, since we define the expected options type in the function signature, preventing any external arguments being used in our code:

```go
func (coll *Collection) InsertOne(ctx context.Context, document interface{},
	opts ...TOptions[options.InsertOneArgs]) (*InsertOneResult, error) {
```

---

I don't anticipate users ever needing to directly interact with `mongo.Options`. IIUC strongly typing the the arg setter parameters will not cause issues, and will result in the same behavior as `any`. If you feel strongly about using `any` I'm happy to make that change.

Reply:
Given the examples you provided which don't require generics or copying `mongoutil.Args`, I agree it doesn't block any known use cases. If a user did want to define a generic method for converting an `*Args` into an `Options`, defining a type constraint that is a subset of `mongoutil.Args` is somewhat inconvenient, but also not blocking. Ideally, if we think that's a use case that many users are interested in, we can provide that utility in the `options` package in the future.

Consider this resolved.